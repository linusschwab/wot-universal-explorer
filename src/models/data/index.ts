export {Data} from './Data'
export {InputData} from './InputData'
export {OutputData} from './OutputData'
