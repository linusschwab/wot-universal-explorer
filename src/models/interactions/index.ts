export {InteractionData} from './InteractionData';
export {InteractionPattern} from './InteractionPattern';
export {Action} from './Action';
export {Event} from './Event';
export {Property} from './Property';
export {ISubscriber} from './ISubscriber';
