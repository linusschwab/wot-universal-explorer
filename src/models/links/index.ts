export {Link} from './Link';
export {HTTPLink} from './HTTPLink';
export {COAPLink} from './COAPLink';
export {Operation} from './Operation';
