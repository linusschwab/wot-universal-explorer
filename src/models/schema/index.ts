export {DataSchema} from './DataSchema'
export {InputSchema} from './InputSchema'
export {OutputSchema} from './OutputSchema'
