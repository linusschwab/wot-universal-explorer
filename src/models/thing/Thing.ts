import {Action, InteractionPattern, Property} from "../interaction";
import {InteractionError} from "../../tools/errors";


export class Thing {

    public name: string;
    public base: string;
    public interaction: InteractionPattern[];

    constructor(name: string, base = '') {
        this.name = name;
        this.base = base;
        this.interaction = [];
    }

    public async readProperty(name: string): Promise<any> {
        let interaction = this.getInteraction(name);

        if (interaction instanceof Property) {
            return interaction.read();
        } else {
            throw new InteractionError('No property with name ' + name);
        }
    }

    public async writeProperty(name: string, data: any): Promise<any> {
        let interaction = this.getInteraction(name);

        if (interaction instanceof Property) {
            return interaction.write(data);
        } else {
            throw new InteractionError('No property with name ' + name);
        }
    }

    public async invokeAction(name: string, data: any = null): Promise<any> {
        let interaction = this.getInteraction(name);

        if (interaction instanceof Action) {
            return interaction.invoke(data);
        } else {
            throw new InteractionError('No action with name ' + name);
        }
    }

    public async subcribeToEvent(name: string): Promise<any> {
        // TODO: Implement
    }

    public registerInteraction(interaction: InteractionPattern) {
        interaction.thing = this;
        interaction.base = this.base;
        this.interaction.push(interaction);
    }

    public getInteraction(name: string): InteractionPattern {
        for (let interaction of this.interaction) {
            if (interaction.name == name) {
                return interaction;
            }
        }
        return null;
    }

    get links() {
        let links = [];

        for (let interaction of this.interaction) {
            for (let link of interaction.links) {
                links.push(link);
            }
        }

        return links;
    }

    public toString() {
        return this.name;
    }
}
