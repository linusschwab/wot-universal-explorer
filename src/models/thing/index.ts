export {Thing} from './Thing';
export {ThingsManager} from './ThingsManager';
