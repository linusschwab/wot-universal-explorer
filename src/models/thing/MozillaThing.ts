import * as WebSocket from "ws";

import {Thing} from "./Thing";
import {WebSocketManager} from "../../controllers";
import {InteractionData} from "../interactions";


export class MozillaThing extends Thing {

    public ws: WebSocket;

    constructor(name: string, type: string, base: string) {
        super(name, type, base);
        this.setupWebSocket();
    }

    private setupWebSocket() {
        let wsUrl = this.base.replace('http://', 'ws://');
        if (this.base.includes(process.env.MOZ_BASE)) {
            wsUrl += '?jwt=' + process.env.MOZ_AUTH;
        }

        // TODO: Handle error(s)
        this.ws = new WebSocket(wsUrl);

        this.ws.on('message', data => this.wsHandleMessage(data));
    }

    public wsHandleMessage(data: WebSocket.Data) {
        let message: any;
        try {
            message = WebSocketManager.parseMessage(data);
        } catch (e) {
            return;
        }

        if (message.messageType === 'propertyStatus') {
            for (let propertyName in message.data) {
                try {
                    const property = this.getProperty(propertyName);
                    const data = new InteractionData(message.data[propertyName]);
                    property.update(data);
                } catch (e) {
                    // Do nothing
                }
            }
        }
    }
}
