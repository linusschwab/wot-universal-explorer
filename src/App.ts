import * as Koa from "koa";
import * as bodyParser from "koa-bodyparser";
import * as cors from "@koa/cors";
import * as fs from "fs";

import {IndexController, TDController, ThingsController} from "./controllers";
import {ThingsManager} from "./models/thing";
import {OpenAPIEncoder, TDParser} from "./tools";


export class App {

    public koa: Koa;
    public things: ThingsManager;

    public static run(): Koa {
        let instance = new App();
        return instance.koa;
    }

    constructor() {
        this.koa = new Koa();
        this.things = new ThingsManager();

        // Mount logger
        this.koa.use(async (ctx, next) => {
            const start = Date.now();
            await next();
            const ms = Date.now() - start;
            console.log(`${ctx.method} ${ctx.url} - ${ms}ms`);
        });

        // Mount body parser
        this.koa.use(bodyParser({strict: false}));

        // Enable CORS
        this.koa.use(cors());

        // Set defaults
        this.koa.use(async (ctx, next) => {
            ctx.set('Content-Type', 'application/json');
            await next();
        });

        // Create routes
        this.routes();

        // Import TDs in folder
        this.importTD();
    }

    private routes() {
        const td = new TDController(this.things);
        this.koa.use(td.router.routes());

        const things = new ThingsController(this.things);
        this.koa.use(things.router.routes());

        // Must be defined last (generic route to serve static files)
        const index = new IndexController();
        this.koa.use(index.router.routes());
    }

    private importTD() {
        // Parse td files to things
        const tdPath = '../public/td/';
        fs.readdirSync(tdPath).forEach(file => {
            let td = fs.readFileSync(tdPath + file, 'utf8');
            let thing = TDParser.parse(td);
            this.things.addThing(thing);
        });

        // Generate openapi file from things
        let json = OpenAPIEncoder.encode(this.things);
        fs.writeFile('../public/swagger-ui/openapi.json', json, 'utf8', (err) => {
            if (err) {
                console.log(err.message);
            }
        });
    }
}
