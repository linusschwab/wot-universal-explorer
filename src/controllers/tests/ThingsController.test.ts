import * as Koa from "koa";
import * as request from "supertest";
import {App} from "../../App";
import {Thing, ThingsManager} from "../../models/thing";


// Create mocks
console.log = jest.fn();
jest.mock('../../models/thing/Thing');
jest.mock('../../models/thing/ThingsManager');

const mockResponse = 10;
(<any>Thing).mockImplementation(() => {
    return {
        readProperty: () => {return mockResponse},
        writeProperty: () => {return mockResponse},
        invokeAction: () => {return mockResponse}
    };
});

const mockThing = new Thing('testthing', 'Thing');
(<any>ThingsManager).mockImplementation(() => {
    return {
        getThing: () => {return mockThing}
    };
});


let app: App;

beforeEach(() => {
    app = new App();
    jest.clearAllMocks();
});

test('get property returns thing response', async () => {
    const response = await request(app.koa.callback())
        .get('/things/testthing/properties/testproperty');

    expect(response.status).toBe(200);
    expect(response.type).toBe('application/json');
    expect(response.body).toBe(mockResponse);
});

test('put property returns thing response', async () => {
    const response = await request(app.koa.callback())
        .put('/things/testthing/properties/testproperty');

    expect(response.status).toBe(200);
    expect(response.type).toBe('application/json');
    expect(response.body).toBe(mockResponse);
});

test('post action returns thing response', async () => {
    const response = await request(app.koa.callback())
        .post('/things/testthing/actions/testaction');

    expect(response.status).toBe(200);
    expect(response.type).toBe('application/json');
    expect(response.body).toBe(mockResponse);
});
