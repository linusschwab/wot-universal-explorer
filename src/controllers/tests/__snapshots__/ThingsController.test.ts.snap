// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`websocket interactions get multiple properties 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testProperty",
    ],
    Array [
      "testProperty2",
    ],
  ],
}
`;

exports[`websocket interactions get property 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testProperty",
    ],
  ],
}
`;

exports[`websocket interactions request action 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAction",
      Object {},
    ],
  ],
}
`;

exports[`websocket interactions request multiple actions 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testAction",
      Object {},
    ],
    Array [
      "testAction2",
      "data",
    ],
  ],
}
`;

exports[`websocket interactions set multiple properties 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testProperty",
      "10",
    ],
    Array [
      "testProperty2",
      "20",
    ],
  ],
}
`;

exports[`websocket interactions set multiple properties 2`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testProperty",
      "10",
    ],
    Array [
      "testProperty2",
      "20",
    ],
  ],
}
`;

exports[`websocket interactions set property 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "testProperty",
      "10",
    ],
  ],
}
`;

exports[`websocket subscriptions subscribe to all 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"subscribe\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Subscribed to all interactions\\"}}",
    ],
  ],
}
`;

exports[`websocket subscriptions subscribe to interaction 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"addSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Subscribed to property testProperty\\"}}",
    ],
  ],
}
`;

exports[`websocket subscriptions subscribe to multiple interactions 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"addSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Subscribed to property testProperty1\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"addSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Subscribed to property testProperty2\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"addSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Subscribed to action testAction\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"addSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Subscribed to event testEvent1\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"addSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Subscribed to event testEvent2\\"}}",
    ],
  ],
}
`;

exports[`websocket subscriptions unsubscribe from all 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"unsubscribe\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Unsubscribed from all interactions\\"}}",
    ],
  ],
}
`;

exports[`websocket subscriptions unsubscribe from interaction 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"removeSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Unsubscribed from property testProperty\\"}}",
    ],
  ],
}
`;

exports[`websocket subscriptions unsubscribe from multiple interactions 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "{\\"messageType\\":\\"removeSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Unsubscribed from property testProperty1\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"removeSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Unsubscribed from property testProperty2\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"removeSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Unsubscribed from action testAction1\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"removeSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Unsubscribed from action testAction2\\"}}",
    ],
    Array [
      "{\\"messageType\\":\\"removeSubscription\\",\\"data\\":{\\"status\\":\\"200\\",\\"message\\":\\"Unsubscribed from event testEvent\\"}}",
    ],
  ],
}
`;
